subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java

```diff
@@ -16,102 +16,115 @@
-        lazyAttributes.keySet().forEach(key -> sorted.put(key, lazyAttributes.get(key).toString()));
-        attributes.keySet().forEach(key -> sorted.put(key, attributes.get(key).toString()));
+        sorted.putAll(getRealizedMap());
         return sorted.toString();
     }
 
     @Override
     public Set<Attribute<?>> keySet() {
-        // Need to copy the result since if the user calls getAttribute() while iterating over the returned set,
-        // realizing a lazy attribute will add to the eager `attributes` map and remove from the `lazyAttributes`.
-        // This avoids a ConcurrentModificationException.
-        return ImmutableSet.copyOf(Sets.union(attributes.keySet(), lazyAttributes.keySet()));
+        return getRealizedMap().keySet();
     }
 
     @Override
     public <T> AttributeContainer attribute(Attribute<T> key, T value) {
         checkInsertionAllowed(key);
-        doInsertion(key, value);
-        return this;
-    }
-
-    private <T> void doInsertion(Attribute<T> key, T value) {
         assertAttributeValueIsNotNull(value);
         assertAttributeTypeIsValid(value.getClass(), key);
-        immutableValue = null;
-        attributes.put(key, attributesFactory.isolate(value));
-        removeLazyAttributeIfPresent(key);
-    }
-
-    private <T> void removeLazyAttributeIfPresent(Attribute<T> key) {
-        lazyAttributes.remove(key);
+        state.put(key, attributesFactory.isolate(value));
+        return this;
     }
 
     @Override
     public <T> AttributeContainer attributeProvider(Attribute<T> key, Provider<? extends T> provider) {
         checkInsertionAllowed(key);
         assertAttributeValueIsNotNull(provider);
-        // We can only sometimes check the type of the provider ahead of time.
-        // When realizing this provider and inserting its value into the container, we still
-        // check the value type is appropriate. see doInsertion
-        if (provider instanceof ProviderInternal) {
-            Class<T> valueType = Cast.<ProviderInternal<T>>uncheckedCast(provider).getType();
-            if (valueType != null) {
-                assertAttributeTypeIsValid(valueType, key);
-            }
+
+        ProviderInternal<T> providerInternal = Cast.uncheckedCast(provider);
+
+        MappingProvider<Isolatable<T>, T> isolated;
+        Class<T> valueType = providerInternal.getType();
+        if (valueType != null) {
+            // We can only sometimes check the type of the provider ahead of time.
+            assertAttributeTypeIsValid(valueType, key);
+            isolated = new MappingProvider<>(null, providerInternal, attributesFactory::isolate);
+        } else {
+            // Otherwise, check the type when the value is realized.
+            isolated = new MappingProvider<>(null, providerInternal, t -> {
+                assertAttributeTypeIsValid(t.getClass(), key);
+                return attributesFactory.isolate(t);
```

Comment:
💭 I know this is just a reimplementation of the existing behavior IRT late checking of the types on lazy attributes, but do we have a test to ensure this still works?

Reply:
That would be the test `adding mismatched attribute types fails when retrieving the key when the provider does not know the type`